Desktop/delete datasets

BryonLewis opened the pull request:
Fixes #593 

- Reorganizes the Recents list into v-row/v-col with spacing and  columns.
- Adds in header for Recents
- Adds in Last Access which is written to each time a user opens data, it now sorts by last access as well.  We use `setRecents` when loading anyways I just added a lastAccess time to each cached metadata.
- Added in Delete icon with confirmation which will delete the project folder as well as remove it from the recents.
- Changed the opening functionality from direct router-link to verifying the data before changing the route.  It does a basic look for the metadata folder/file and check to make sure the source Dir exists.  If not it will error out and ask the user if they want to delete the dataset.  I did it this way because I didn't want to mess around with some additional error handling in the common Viewer.vue and this seemed like a quick verification way to see if the folders/data exists without processing too much.

![image](https://user-images.githubusercontent.com/61746913/126164785-d2f0cb38-ed01-428e-a728-891338a9445c.png)

subdavis commented:
I sorta wasn't expecting a redesign of the recents list to be part of this issue.  I've redesigned this list myself several times and never liked my results enough to propose the change.  Could you maybe add your thoughts on how this change improves usability? The current list is intentionally minimal, so I'd like to understand the justification for moving away from that.

* Vuetify wastes a ton of vertical space.  This change reduces information density of the list to about a third, which requires more scrolling to find what you want.  If you can see less information, you have to scroll more, which I find annoying.
* I don't think table layout is ideal for this page because requiring vertical alignment of sections wastes a lot of horizontal space.  In your screenshot, there are a lot of unnecessary line wraps.
* I'm not sure featuring the the last access time prominently like this  is useful for an end user.  It's good for us to have to order the list by, but I'm not yet convinced that it's valuable to show it.

## Possible alternative 1

Without changing the list to a table, switch to a flex row and add a `v-spacer` between location and last access.  Show the icon only on row hover just like how the edit button appears in the type list.  This could either be at the right side or the left side.  You could even replace the image or video icon with an `X` on hover instead of using the trashcan?

```
[icon] Stereo /home/local/KHQ/bryon.lewis/dev/whatever/foldername                 7/16/2021, 8:41 AM
[icon] Lehi /home/local/KHQ/bryon.lewis/dev/whatever/lehidata                     7/16/2021, 8:41 AM
[icon] Fish Training Data2 /home/local/KHQ/bryon.lewis/dev/whatever/folder        7/16/2021, 8:41 AM
```

## Possible alternative 2

Keep the table layout, but remove the vertical dividers and/or make the padding the same as it was before.

## Other comments

* Change the alignment on the delete button such that it's right aligned with the end of the table and not touching the edge of the container card.
* Remove the seconds from last access time. It makes the timestamp harder to read at a glance IMO.

subdavis commented:
Also, FWIW, #593 doesn't request design changes to the recents list, so I assume you did this mostly because you felt that the new feature (delete button) that you added to the list pushed the existing design past what it could accommodate and it needed to change.  

We ran into the same situation when you added the import button for CSV, and I liked that your approach was to add the feature as it was written and consider the design implications (needing to redesign that area of the toolbar) as out of scope.  Perhaps the same approach applies here?

BryonLewis commented:
> Could you maybe add your thoughts on how this change improves usability? The current list is intentionally minimal, so I'd like to understand the justification for moving away from that.

I felt that the design changes were necessary to get the delete icon into a consistent location for usability purposes.  I could add them to left side and that would place them in the same spot all time.  For some reason left side delete icons just feel weird to me.

- One initial thing I thought about when editing the view was hiding/condensing the full file path location of the source media.  It seems to me the final piece of information I would use if I was distinguishing between data and takes up a lot of horizontal space.
- I didn't like how before the non-alignment distances between pieces of information in the list.  They way that the source media location could be offset with really short dataset names or the fact that if there is a really long source media path or a really long dataset name you are going to get wrapping anyways which is going to cause some weird instances.  I know the below contains extremely exaggerated corner cases, but that could occur in the case of network mounts.
![image](https://user-images.githubusercontent.com/61746913/126181175-352401b1-32f6-4734-b6d7-d4031974a645.png) 
![image](https://user-images.githubusercontent.com/61746913/126184660-a68c4605-13dc-4b11-972d-a348f70d5cf9.png)
These second shots are a misleading, we may want to specify a minimum width like we do with the annotator.

- Having datasets in columns that wraps I think requires having some sort of divisions between the rows.  Especially in the case of right align buttons and short dataset/location names.  I do see your point in that there is some unnecessary padding to the top/bottom of the rows and that should be removed.
- I could see adding in a hover/delete or context icons, but if that occurs I would want the button to always appear in the same spot without being offset based on the name length.
- I agree that the time doesn't need to require seconds.  I do however think that last access time provides some vital contextual information when looking at list of data that may not have the most intuitive names liked `DatasetAACD-Cam24532`

Ideally I would have:
```
[icon] Stereo           7/16/2021, 8:41 AM [info.ico] [delete.ico]
-------------------------------------------------------------------------------
[icon] Lehi             7/16/2021, 8:41 AM [info.ico] [delete.ico]
-------------------------------------------------------------------------------
[icon] Fish Training    7/16/2021, 8:41 AM [info.ico] [delete.ico]
          Data2 
-------------------------------------------------------------------------------
```
`info.ico` creating a dialog and providing more info like the source media, a way to open the project directory and to open the source media folder.

Finally some more exaggerated and relatively implausible instances:
![image](https://user-images.githubusercontent.com/61746913/126186935-a74efbdf-d4a9-4afc-b859-606e8a98e895.png)
![image](https://user-images.githubusercontent.com/61746913/126186955-056d898f-2d5e-4083-aa24-4a8c07d2541b.png)

subdavis commented:
I don't think it makes sense to implement a table layout using flex boxes.  The headers aren't aligned with the columns, and unless you set fixed widths, you can't line them up.  If you do set fixed widths, you end up with a sub-optimal layout because normal html tables derive their widths based on the cell content

> One initial thing I thought about when editing the view was hiding/condensing the full file path location of the source media. It seems to me the final piece of information I would use if I was distinguishing between data and takes up a lot of horizontal space.

I consider path more important than the timestamp, so perhaps we should ask someone.

> I could see adding in a hover/delete or context icons, but if that occurs I would want the button to always appear in the same spot without being offset based on the name length.

Agree

Here's a possible option for going the flex route:

![Screenshot from 2021-07-19 13-25-48](https://user-images.githubusercontent.com/4214172/126201825-7cfb2660-0831-4fd1-aa88-be54afbdcf41.png)
![Screenshot from 2021-07-19 13-25-34](https://user-images.githubusercontent.com/4214172/126201826-64196b84-0c41-49bc-8e67-98d8af510b79.png)

However, if you really prefer a table layout, I think it would be better to just use a `v-data-table` like the training table does.

BryonLewis commented:
Yeah I wondered about using a data-table at some point instead of doing it my way.  I can switch to a flex row and v-spacer to make it work.  I just want to try to maintain some consistency and not make it too difficult if we have to other data elements/icons in here in the future.

BryonLewis commented:
went with the data-table option.  secondary benefit of name/last access sorting as well (I'm assuming they rely on the data-field default sort).  May have to implement a proper sort of the Last access field then.
![image](https://user-images.githubusercontent.com/61746913/126821367-fd90b387-8488-4eae-bb95-057a03116767.png)

aashish24 commented:
Looking good. Just wanted to provide a very minor feedback (not sure if this has been discussed). Instead of "access" perhaps "accessed" or "last accessed" might be more useful word? Google Drive has "Last modified"

f4str commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -262,69 +335,67 @@ export default defineComponent({
                   color="primary lighten-2"
                 >
-                  {{ getTypeIcon(recent) }}
+                  {{ getTypeIcon(item) }}
```
Do you think it's worth adding a hover tooltip saying the dataset type? Just throwing the idea out there, might be something too extra so feel free to ignore.

BryonLewis replied:
I like that because the icons are kind of ????? for representation.

f4str commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -262,69 +335,67 @@ export default defineComponent({
+                  color="error"
+                  icon
+                  @click="confirmDeleteDataset(item.id, item.name)"
+                >
```
If you do go with my suggestion on removing the `Delete` header from the delete button column, it might be worth adding a hover tooltip similar to the viewer:

![Screenshot from 2021-07-26 09-59-50](https://user-images.githubusercontent.com/30541973/127001797-bc007faf-b81f-43b0-9748-a43c8a28add0.png)

Again, just throwing the idea out there. Feel free to ignore if its too extra.

subdavis commented on client/platform/desktop/backend/native/common.ts:
```diff
@@ -604,6 +604,30 @@ async function _initializeProjectDir(settings: Settings, jsonMeta: JsonMeta): Pr
+): Promise<boolean> {
+  // confirm dataset Id exists
+  const projectDirInfo = await getValidatedProjectDir(settings, datasetId);
+  fs.rmdirSync(projectDirInfo.basePath, { recursive: true });
```
Don't use synchronous fs methods.   `await fs.remove(projectDirInfo.basePath)` is an async recursive equivalent.

https://github.com/jprichardson/node-fs-extra/blob/master/docs/remove.md

BryonLewis replied:
I don't know why I did a synchronous delete there but this brings up another question I had.  I feel very awkward about implementing a recursive delete on a user's file system based on a folder.  Should I be more explicit in the files I am deleting from the `projectDir` instead of just doing a recursive delete on the whole folder?

f4str commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -111,29 +123,90 @@ export default defineComponent({
+          confirm: true,
+          positiveButton: 'Delete',
+          negativeButton: 'Cancel',
+        });
```
This is something to note for the future. In the CLI Tools PR, there is a new function to validate datasets. One thing we can do is call that validate function when attempting to open a dataset to check if it's corrupted and notify the user. 

Not something necessarily for this PR, but thought it's worth mentioning.

BryonLewis replied:
The error here could be a missing metadata file, missing project folder or missing media source folder.  I've updated it to to remove the in-message delete option and give instructions to check the error details to fix.  I don't know of all the errors so I think I'm going to leave the full trace of the error to hopefully provide more context.
The one think I am removing is the initial `Error:` so it will be more like `Error invoking remote meth 'check-dataset':` followed by the context/info that caused that error.

subdavis commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -86,18 +88,28 @@ export default defineComponent({
+    async function confirmDeleteDataset(datasetId: string, datasetName: string) {
+      const result = await prompt({
+        title: 'Confirm',
+        text: `Do you want to delete dataset ${datasetName}`,
```
Please include very scary warning about exactly what this will remove.

1) It will not remove source media, such as images or video.
2) It will not remove annotation files that were imported when the dataset was created.
3) It will completely remove any annotations that have been created in DIVE.  

Maybe even mention that you can use the export button to create a backup before deletion.

subdavis commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -111,29 +123,90 @@ export default defineComponent({
+          title: 'Error Loading Data',
+          text: e,
+          confirm: true,
+          positiveButton: 'Delete',
+          negativeButton: 'Cancel',
```
Maybe don't allow someone to delete a dataset from this menu.  Force them thorough the regular delete menu and make them look at the scary warnings

subdavis commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -111,29 +123,90 @@ export default defineComponent({
+    async function preloadCheck(recent: JsonMetaCache) {
+      //Attempts to preload the data to see if there are any isues
+      try {
+        await api.checkDataset(recent.id);
```
The most common error I can think of involves the case where the user changed the location of the source media, which need to be re-located to launch the viewer.  This doesn't check that, right?

BryonLewis replied:
Yeah I didn't know how far to go for that.  It checks to make sure the `originalBasePath` (folder which has the source video/images) exists.  So it goes part of the way to make sure the path exists.  I didn't want to go through and check all of the individual images.

subdavis replied:
I think this is reasonable for this PR.  I'll mark this one resolved.

subdavis commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -111,29 +123,90 @@ export default defineComponent({
+        sortable: false,
+      },
+    ];
+    const toDisplayString = (dateString: string) => new Date(dateString).toLocaleString('en-US',
+      {
+        hour12: false,
+        month: '2-digit',
+        day: '2-digit',
+        year: '2-digit',
+        hour: '2-digit',
+        minute: '2-digit',
+      });
```
we already have the `moment.js` dependency, might as well use it here.  https://momentjs.com/docs/#/displaying/

I think you end up with much nicer control over the output string.

`moment(dateString).format('MM DD YY, H:mm')` or something like that.

subdavis commented on client/platform/desktop/frontend/store/dataset.ts:
```diff
@@ -43,7 +45,7 @@ const datasets = ref({} as Record<string, JsonMetaCache>);
 const recents = computed(() => {
   const list = Object.values(datasets.value)
-    .sort((a, b) => Date.parse(b.createdAt) - Date.parse(a.createdAt));
+    .sort((a, b) => Date.parse(b.accessedAt) - Date.parse(a.accessedAt));
```
Now that you're doing sorting with v-data-table, can you just use a default sort on that table and remove this?

Or do you think it's better to do the default sort here?

BryonLewis replied:
It's one of the things I need to check.  I may want to re implement the `sort` in the table to use the same `Date.parse`, because I imagine it's just sorting on the string name and I'm not sure if that is the best way.

f4str commented on client/platform/desktop/frontend/components/Recent.vue:
```diff
@@ -262,69 +335,67 @@ export default defineComponent({
+              dense
+              v-bind="{ headers: headers, items: filteredRecents }"
+              :footer-props="{ itemsPerPageOptions: [10, 30, -1] }"
+              no-data-text="No data loaded"
```
We need to update the screenshot for desktop on <https://kitware.github.io/dive/Screenshots/#browse-files>

subdavis commented on client/platform/desktop/frontend/store/dataset.ts:
```diff
@@ -83,18 +85,27 @@ function locateDuplicates(meta: JsonMeta) {
  * @param id dataset id path
  */
-function setRecents(meta: JsonMeta) {
+function setRecents(meta: JsonMeta, accessTime?: string) {
```
docstring for this function is out of date, just remove it?

f4str reviewed (changes requested):
This is great! I ran a couple tests and looks like everything works without any major issues. I really like the style and functionality of the data table allowing to sort by name or last accessed data. I also think the prompt to delete the dataset when it does not exist is a good addition. 

The biggest point I wanted to bring up is whether the "Recents" page is even a recents page anymore. Due to the new data table, this has all of the datasets now organized and sorted by access date. My suggestion is to rename the "Recents" to "Datasets" since this is how the "Training" page refers them as. Since it's a trivial change, here's a screenshot of how it looks:

![image](https://user-images.githubusercontent.com/30541973/127006602-2b2db675-fbb8-418f-ba8a-f0ceff53e0b3.png)

However, I am aware that the "Recents" name has been used for a long time so it might be a big change for the user perspective. 

All of my other comments are minor.

subdavis reviewed (changes requested):
Along with the requested changes inline, what do you think about this style suggestion?

``` html
                 <div
                    v-else
                    class="link primary--text text--lighten-3 text-subtitle-1 pt-1"
                    style="line-height: initial;"
                    @click="preloadCheck(item)"
                  >
                    {{ item.name }}
                  </div>
                  <div class="grey--text text-caption">
                    {{ item.originalBasePath }}
                  </div>
```

`text-subtitle-1` and `pt-1` ,  and `line-height: initial;` for the dataset name,

`text-caption` for the path

And force the `Accessed` width to `140` so the timestamp doesn't wrap.

![Screenshot from 2021-07-26 10-42-01](https://user-images.githubusercontent.com/4214172/127008168-c4250e51-77a0-4c49-b0b4-f06f0dfe9c34.png)

BryonLewis commented:
Still need to determine on if we change `Recents` to `Datatsets` as well as take a screenshot for the help.

BryonLewis commented:
Updated screenshot and fixed a small issue with stereo display (due to the way `loadMetadata` works for `multi` datasets).

Decided against renaming 'Recents' to 'Datasets' for now just because it looks like ther are 60+ variables/functions/files that would need renaming.  This can be done in another PR instead of complicating this one.

subdavis reviewed (approved):
LGTM, thanks for all the revisions.

f4str reviewed (approved):
Everything looks good on my end as well.
